certora/specs/ERC20Wrapper.spec

```diff
@@ -0,0 +1,190 @@
+import "helpers.spec"
+import "ERC20.spec"
+
+methods {
+    underlying()                       returns(address) envfree
+    underlyingTotalSupply()            returns(uint256) envfree
+    underlyingBalanceOf(address)       returns(uint256) envfree
+    underlyingAllowanceToThis(address) returns(uint256) envfree
+
+    depositFor(address, uint256)       returns(bool)
+    withdrawTo(address, uint256)       returns(bool)
+    recover(address)                   returns(uint256)
+}
+
+use invariant totalSupplyIsSumOfBalances
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ Helper: consequence of `totalSupplyIsSumOfBalances` applied to underlying                                           │
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+function sumOfUnderlyingBalancesLowerThanUnderlyingSupply(address a, address b) {
+    if (a != b) {
+        require underlyingBalanceOf(a) + underlyingBalanceOf(b) <= underlyingTotalSupply();
+    } else {
+        require underlyingBalanceOf(a) <= underlyingTotalSupply();
+    }
+}
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ Invariant: wrapped token can't be undercollateralized (solvency of the wrapper)                                     │
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+invariant totalSupplyIsSmallerThanUnderlyingBalance()
+    totalSupply() <= underlyingBalanceOf(currentContract)  &&
+    underlyingBalanceOf(currentContract) <= underlyingTotalSupply() &&
+    underlyingTotalSupply() <= max_uint256
+    {
+        preserved {
+            requireInvariant totalSupplyIsSumOfBalances;
+            require underlyingBalanceOf(currentContract) <= underlyingTotalSupply();
```

Comment:
Can't we write:

```suggestion
            requireInvariant totalSupplyIsSmallerThanUnderlyingBalance;
```

Reply:
👍🏻

Reply:
we could do

```sumOfUnderlyingBalancesLowerThanUnderlyingSupply(currentContract, 0)```

Reply:
> It a bit like using a variable inside its own declaration.

No, that is not the case according to the [docs on invariants](https://docs.certora.com/en/latest/docs/cvl/invariants.html#invariants-and-induction). The `preserve` block is included as a precondition, so by using a "recursive" `requireInvariant` in its own preserve block, we're just stating that this is an inductive property.